### PR TITLE
Add DPR Reader

### PR DESCRIPTION
**Pull request by KaiSun314**
Draft for adding a Reader module and DPRReader in PyGaggle (https://github.com/castorini/pygaggle/issues/149).

TODO: Test DPRReader by replicating the results in DPR paper.

**Description of Code**

pygaggle/model/evaluate.py: class ReaderEvaluator - Takes in the examples (query, texts, ground truth answers), predicts answers using the DPRReader, and compares the top answer with the ground truth answers using exact match. Like https://github.com/facebookresearch/DPR/blob/f403c3b3e179e53c0fe68a0718d5dc25371fe5df/train_reader.py#L186.

pygaggle/reader/dpr_reader.py: class DensePassageRetrieverReader - Predicts answers from a query and texts, like in https://huggingface.co/transformers/v3.2.0/_modules/transformers/tokenization_dpr.html.

pygaggle/run/evaluate_passage_reader.py: The main content starts on Line 82, which will group the examples into 1000-sized batches (since 8757 queries at once gave me RAM issues) and evaluate them. The exact match scores are extended into one list, and averaged to get the overall EM score.

Colab link: https://colab.research.google.com/drive/1d3BYhpo71YUKcCOLyPLT4Ulvc71yxF78?usp=sharing

### CODE REVIEW

**Review by MXueguang — COMMENTED** [review]
LGTM!

**Comment by MXueguang** [comment]
to evaluate the answer prediction, you need to implement `exact match` (EM).
make it identical to what DPR did here https://github.com/facebookresearch/DPR/blob/f403c3b3e179e53c0fe68a0718d5dc25371fe5df/dpr/data/qa_validation.py#L140

File: pygaggle/reader/dpr_reader.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+from .base import Reader, Answer
+from ..reranker.base import Query, Text
+
+class DPRReader(Reader):
+    def __init__(
+        self,
+        model_name: str,
+        tokenizer_name: str = None,
+        num_spans: int = 16,
```
Comment (ronakice) [inline pygaggle/reader/dpr_reader.py:13]: can you add a bit of a description for all the function parameters, some of these might not be 100% apparent to non-QA folks

**Comment by MXueguang** [comment]
LGTM, 
I think we can merge this version since the functionalities are ready and @KaiSun314 is able to replicate the results.

There are some additions/modifications needed, but I think we should leave that for the next iteration.
e.g.
1. adjustable batch-size
2. integrate with `pyserini`
3. e2e QA

CC @lintool @ronakice

**Comment by ronakice** [comment]
Once we have a colab link to replicate on some subset, we can merge!

**Review by MXueguang — APPROVED** [review]

**Comment by KaiSun314** [comment]
@ronakice The colab link: https://colab.research.google.com/drive/1d3BYhpo71YUKcCOLyPLT4Ulvc71yxF78?usp=sharing

**Comment by ronakice** [comment]
Since this is working for now I'll merge this in and then we can make updates in some PRs! Great job @KaiSun314 !

**Pull request merged**